### PR TITLE
Changing the rest factory for UserSync

### PR DESCRIPTION
**Pull request by ruma-nair**
JIRA ticket:

- [ ] Approved by a PI tech lead

### CODE REVIEW

File: src/test/java/com/appdirect/sdk/web/exception/UserSyncApiExceptionHandlerTest.java [hunk 1]
```diff
@@ -1,173 +0,0 @@
...
-import org.springframework.http.client.ClientHttpResponse;
-
-import com.appdirect.sdk.exception.UserSyncException;
-import com.appdirect.sdk.exception.UserSyncTooManyRequestsException;
-
-public class UserSyncApiExceptionHandlerTest {
-	private UserSyncApiExceptionHandler userSyncApiExceptionHandler;
-
-	@Before
-	public void setup() throws Exception {
```
Comment (ruma-nair) [inline src/test/java/com/appdirect/sdk/web/exception/UserSyncApiExceptionHandlerTest.java]: Addressed in the UserSyncClientTest

File: src/test/java/com/appdirect/sdk/appmarket/usersync/UserSyncApiClientTest.java [hunk 2]
```diff
@@ -14,75 +14,154 @@
...
-	private static final String BASE_URL = "https://orchardbox.appdirect.com";
 	private static final String OAUTH_KEY = "testKey";
 	private static final String OAUTH_SECRET = "testSecret";
+	private static final String USER_SYNC_ENDPOINT = "/api/sync/v1/tasks";
+	private static final String HOST = "localhost";
+	private String hostUrl;
 	private SyncedUser syncedUser;
 
+	@Rule
+	public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().dynamicPort());
```
Comment (EmilDafinov) [inline src/test/java/com/appdirect/sdk/appmarket/usersync/UserSyncApiClientTest.java]: 👍 for the dynamic port

File: src/test/java/com/appdirect/sdk/appmarket/usersync/UserSyncApiClientTest.java [hunk 3]
```diff
@@ -94,14 +173,25 @@ private SyncedUser createSyncedUser() {
...
-	private UserSyncRequestPayload createExpectedPayload(SyncedUser syncedUser, UserSyncRequestPayloadOperation operation) {
-		return UserSyncRequestPayload.builder()
-				.accountIdentifier(syncedUser.getAccountIdentifier())
-				.developerIdentifier(syncedUser.getDeveloperIdentifier())
-				.email(syncedUser.getEmail())
-				.type("ASSIGNMENT")
-				.userIdentifier(syncedUser.getUserIdentifier())
-				.operation(operation.toString())
-				.build();
+	private String getAssignRequestBody() {
```
Comment (EmilDafinov) [inline src/test/java/com/appdirect/sdk/appmarket/usersync/UserSyncApiClientTest.java]: Could we extract the expected payloads into JSON files and load them as resources?

File: src/main/java/com/appdirect/sdk/web/oauth/OAuthRestTemplateFactory.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.appdirect.sdk.web.oauth;
+
+import org.springframework.web.client.RestTemplate;
+
+public interface OAuthRestTemplateFactory {
```
Comment (EmilDafinov) [inline src/main/java/com/appdirect/sdk/web/oauth/OAuthRestTemplateFactory.java]: Seems to be doing exactly the same thing as the `RestOperationsFactory`
Perhaps we can remove the `RestOperationsFactory` and rely on implementations of this interface instead?
That way the default and the UserSync implementations would be in charge of setting the correct error handler

Comment (ruma-nair) [inline src/main/java/com/appdirect/sdk/web/oauth/OAuthRestTemplateFactory.java]: Agreed.

File: src/test/java/com/appdirect/sdk/appmarket/usersync/UserSyncApiClientTest.java (hunk 3 above)
Comment (ruma-nair) [inline src/test/java/com/appdirect/sdk/appmarket/usersync/UserSyncApiClientTest.java]: Sure!

File: src/main/java/com/appdirect/sdk/web/oauth/OAuthRestTemplateFactory.java (hunk 4 above)
Comment (ian-beckles) [inline src/main/java/com/appdirect/sdk/web/oauth/OAuthRestTemplateFactory.java]: I agree on these being similar but to me it seems that the interface should be RestOperationsFactory or something to that effect. I think in the future it is certainly possible that we would want to request a RestTemplate where we don't need OAuth. Also maybe the method should be RestTemplate getOAuthRestTemplate(String key, String secret);

File: src/main/java/com/appdirect/sdk/web/oauth/UserSyncRestTemplateFactoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	private static final int DEFAULT_CONNECT_TIMEOUT = 10000;
+	private static final int DEFAULT_READ_TIMEOUT = 60000;
+
+	@Override
+	public RestTemplate getOAuthRestTemplate(String key, String secret) {
+		BaseProtectedResourceDetails oauthCredentials = new BaseProtectedResourceDetails();
+		oauthCredentials.setConsumerKey(key);
+		oauthCredentials.setSharedSecret(new SharedConsumerSecretImpl(secret));
+
+		HttpComponentsClientHttpRequestFactory clientHttpRequestFactory = new HttpComponentsClientHttpRequestFactory();
```
Comment (ian-beckles) [inline src/main/java/com/appdirect/sdk/web/oauth/UserSyncRestTemplateFactoryImpl.java]: Do we really need to create a new HttpComponentsClientHttpRequestFactory for each RestTemplate that we return?

**Comment by ian-beckles** [comment]
minor: RestTemplateFactory et. al probably shouldn't be in the OAuth package

File: src/test/java/com/appdirect/sdk/web/oauth/DefaultRestTemplateFactoryTest.java [hunk 6]
```diff
@@ -9,45 +9,32 @@
...
-		String testKey = "testKey";
-		String testSecret = "testSecret";
-
-		//When
-		RestTemplate restClient = testRestOperationsFactory.restOperationsForProfile(testKey, testSecret);
-
-		//Then
-		assertThat(restClient.getErrorHandler()).isEqualTo(appmarketEventClientExceptionHandler);
+	public void testGetRestTemplate_ReturnsOauthSignedRequestFactory() throws Exception {
+		RestTemplate restTemplate = defaultOAuthRestTemplateFactoryImpl.getOAuthRestTemplate("some-key", "some-secret");
```
Comment (EmilDafinov) [inline src/test/java/com/appdirect/sdk/web/oauth/DefaultRestTemplateFactoryTest.java]: Could we have the Given / When /Then ?

File: src/test/java/com/appdirect/sdk/web/oauth/UserSyncRestTemplateFactoryImplTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	public void testGetRestTemplate_ReturnsOauthTemplate() throws Exception {
+		//When
+		RestTemplate restTemplate = userSyncRestTemplateFactory.getOAuthRestTemplate("some-key", "some-secret");
+
+		//Then
+		assertThat(restTemplate).isInstanceOf(OAuthRestTemplate.class);
+		OAuthRestTemplate oauthRestTemplate = (OAuthRestTemplate) restTemplate;
+
+		ProtectedResourceDetails resource = oauthRestTemplate.getResource();
+		assertThat(resource.getConsumerKey()).isEqualTo("some-key");
```
Comment (EmilDafinov) [inline src/test/java/com/appdirect/sdk/web/oauth/UserSyncRestTemplateFactoryImplTest.java]: //Then

**Pull request merged**